File: 2022/Python/02.py
def generator(input) :
    duels = []
    for duel in input.splitlines():
        [a, b] = duel.split(' ')
        duels.append([ord(a) - 63 , ord(b) - 88])
    return duels

def part_1(input) :
    return solver(input , real_rules = False)

def part_2(input) :
    return solver(input , real_rules = True)

def solver(input, real_rules) :
    s = 0
    for [a, b] in input :
        v = [(b + (2 - real_rules) * a) % 3 , b]
        s += 1 + 3 * v[real_rules] + v[1 - real_rules]
    return s

File: 2022/Rust/src/bin/02.rs
aoc2022::main!();

fn generator(input : &str) -> Vec<(u32, u32)> {
    input  
        .lines()
        .map(|l| (l.chars().nth(0).unwrap(),l.chars().nth(2).unwrap()))
        .map(|l| (l.0 as u32 - 63 , l.1 as u32 - 88))
        .collect_vec()
}

fn part_1(input : Vec<(u32, u32)>) -> u32 {
    solver(input, false)
}

fn part_2(input :Vec<(u32, u32)>) -> u32 {
    solver(input, true)
}

fn solver(input : Vec<(u32, u32)>, real_rules : bool) -> u32 {
    input
    .iter()
    .map(|l| [(l.1 + (1 + real_rules as u32) * 2 * l.0) % 3 , l.1])
    .map(|l| 3 * l[real_rules as usize] + l[!real_rules as usize] + 1)
    .sum()
}
